[WIP] Azure Blob backend state provider

jjcollinge opened the pull request:
**This is WIP and I've submitted it as a PR to get early feedback on the design before I've invested too much time. I appreciate it needs a little more love :)**

### What this PR does?
This PR attempts to abstract out a blob interface from the current `filestate` local backend to allow other blob backends such as Azure Blob Storage, S3, etc. to implement it. I have provided an Azure Blob Storage implementation which does exactly this.

### Status
The code currently supports both local file backend and Azure Blob Storage backend. I've smoked tested both of these implementations but still need to formalise some unit tests and integration tests. I'm looking for some initial feedback and discussion on whether this is a suitable method for extending the current Pulumi backends. I have only spent a day working with the codebase so far so I may have missed some functionality, please let me know and I'll look into it! As this is WIP I'll likely continue to push commits to it.

### Usage
You can use the local file backend as per normal, using `pulumi login --local`.
You can use the Azure Blob Storage backend using the following Cloud URL format:
`pulumi login --cloud-url azure://my-storage-account.blob.core.windows.net/container-name`. Currently, you can pass the associated storage account key as an environment variable `AZURE_STORAGE_ACCOUNT_KEY` or enter the value when prompted by the CLI.
**NOTE:** The storage account must exist but the container is not required as it will be created dynamically. I have kept the full Azure Blob Storage path in the URL to support all Azure environments (public, Government and China).

### TODO
- [x] File/Azure filestate implementations
- [x] Tests
- [x] Access token handling
- [x] State locking/leasing
- [ ] Update docs

### Out of scope
- More login options
- Go-cloud for AWS, GCP
- Key management (Key Vault)

### Related issues:
#1941 
#1966

ellismg commented on pkg/backend/blobstate/storage/azure/bucket.go:
```diff
@@ -0,0 +1,173 @@
+		return nil, err
+	}
+	blob, err := ioutil.ReadAll(res.Body())
+	if err != nil {
```
Do we need to close the body after reading it?

jjcollinge replied:
Yes, good catch - I'll add a deferred close

ellismg commented:
Thanks for opening this @jjcollinge and sorry it took us so long to get back to you.

Overall this looks like it's trending in the right direction.  I think we'd prefer if we kept the `filestate` and `local` terminology instead of moving everything over to `blob`.  For your key-vault TODO, what were you thinking?  Is this to get a access key for the storage bucket from KeyValut itself?

jjcollinge commented:
@ellismg thanks for the review - I'll change back to your preferred terminology. As far as the key vault  TODO was concerned, it was more a reminder for me to investigate whether we could use key vault to Encrypt/Decrypt values. Azure Storage has [preview support for RBAC](https://docs.microsoft.com/en-us/azure/storage/common/storage-auth-aad-rbac?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json) so I'd switch the current storage account key for a service principal `clientId:clientSecret` with fine-grained RBAC permissions - I'd then be able to read/write to blob and encrypt/decrypt secrets using an externally managed key in key vault rather than relying on a shared passphrase. I haven't really looked into this yet to check it makes sense and it's probably outside the scope of the initial PR.

jjcollinge commented:
I updated the naming so that the package is named `filestate` again, I renamed the concept of `localBackend` and `localStack` etc. to `fileBackend` and `fileStack` that can then use either a `local` storage provider or `azure` storage provider. I think this makes more sense than keeping it as a `localBackend` and `localStack` which can then use _remote_ storage providers like `azure`...

ellismg commented:
>  I think this makes more sense than keeping it as a localBackend and localStack which can then use remote storage providers like azure...

Yeah, sounds reasonable.

The additional context stuff is still confusing to me, to be honest.  I think that you have to do this because the azure go sdk requires a context to be passed.  I am wondering if it makes more sense to just use context.Background() or context.TODO() explicitly in your implementation of the provider, and save us the trouble of threading everything through as part of this.

ellismg commented on pkg/backend/filestate/testdata/Pulumi.yaml:
```diff
@@ -0,0 +1,3 @@
+name: test-project
```
Do we need this test project today?

jjcollinge replied:
It's currently used by the tests in `pkg/backend/filestate/backend_test.go`

ellismg commented on pkg/backend/filestate/storage/azure/storage.go:
```diff
@@ -0,0 +1,36 @@
+		fmt.Printf("Using access token from %s\n", accessTokenEnvVar)
+	} else {
+		// TODO: Support other login modes
+		return errors.Errorf("Please login to your Azure backend using `pulumi login`")
```
I'm not 100% sure this is an actionable error.  Presumably running `pulumi login` is going to fail, because it will just come back here and the envvar for the access token won't be set.  Should be be reading the token here interactively?

jjcollinge replied:
Yeh sorry this was just a placeholder I left in before I implemented the prompt - I've added it in the latest commit.

jjcollinge commented:
> 
> 
> > I think this makes more sense than keeping it as a localBackend and localStack which can then use remote storage providers like azure...
> 
> Yeah, sounds reasonable.
> 
> The additional context stuff is still confusing to me, to be honest. I think that you have to do this because the azure go sdk requires a context to be passed. I am wondering if it makes more sense to just use context.Background() or context.TODO() explicitly in your implementation of the provider, and save us the trouble of threading everything through as part of this.

I'm happy to define it in-line in the provider. My original thinking was to allow you to define global context properties (timeout/cancellations) at a top level that you could propagate to the implementations rather than having different rules for each implementation. But on reflection, I think you're right and the contexts should be implementation specific.

jjcollinge commented:
> 
> 
> > I think this makes more sense than keeping it as a localBackend and localStack which can then use remote storage providers like azure...
> 
> Yeah, sounds reasonable.
> 
> The additional context stuff is still confusing to me, to be honest. I think that you have to do this because the azure go sdk requires a context to be passed. I am wondering if it makes more sense to just use context.Background() or context.TODO() explicitly in your implementation of the provider, and save us the trouble of threading everything through as part of this.

It's probably my lack of understanding that's causing the confusion - your existing backend interface methods pass through a context object - this is used by the client in the `httpstate` backend but is currently not used in the `filestate` backend as local IO doesn't require a context. Since we're following a similar pattern to `httpstate` and adding a `bucket` (comparable to a client) to the `filestate` backend which most implementations will make remote calls (except `local`) I didn't see any harm in passing the context down to the individual bucket and letting the `local` bucket drop it but other implementations such as `azure` use it. Have I misunderstood?

jjcollinge commented:
@ellismg I think this now includes the basic features I wanted to enable in this PR. I'll not make any more commits until I've had feedback of you or others at Pulumi to stop you chasing a moving target.

ellismg commented:
> I'll not make any more commits until I've had feedback from you or others at Pulumi to stop you chasing a moving target.

Thanks!  Do you mind rebasing or merging in master to deal with the conflicts?  I'll take a pass tonight.

jjcollinge commented:
@ellismg - I had to tweak the existing CI a bit to make sure it ran the new backend tests I've added correctly. These tests are integration tests but I didn't want to fold them into your existing black box integration tests so I've kept them separate. Most run as part of `test_fast` whilst the Update tests run as part of `test_all`. If you want to enable the Azure backend tests to run as part of CI, you simply need to supply the environment variables `AZURE_URL` and `AZURE_STORAGE_ACCOUNT_KEY`, the tests will then pick these up and run against the Azure and local provider. There are a few outstanding //TODO comments in the code where I've asked questions - I'll inline them here:
https://github.com/pulumi/pulumi/blob/d61fc7253567b1cfd4c8d148d2d26a8211e5e534/pkg/backend/filestate/storage/local/bucket.go#L122
https://github.com/pulumi/pulumi/blob/d61fc7253567b1cfd4c8d148d2d26a8211e5e534/tests/integration/integration_test.go#L196
Another point I'd like to highlight is the `locking` test. This unfortunately relies on timings, which is variable and different for each backend provider. I'm not sure of a better method to test the locking semantics work as expected so I'm open to suggestions: https://github.com/pulumi/pulumi/blob/d61fc7253567b1cfd4c8d148d2d26a8211e5e534/pkg/backend/filestate/backend_test.go#L468

jjcollinge commented:
@ellismg - have you had chance to take a look through the PR yet? Appreciate you're super busy just I don't want to forget too much context before I have to make further changes.

oliverholliday commented:
This looks great and definitely something we'd like to start using ASAP.

joeduffy commented:
@CyrusNajmabadi Now that https://github.com/pulumi/pulumi/pull/2455 has landed, should this PR be closed out?

CyrusNajmabadi commented:
Yup yup.  This shoudl work now with #2455.  Please try it out in our next release and let us know if you run into any issues.
